docs/ide/api-designs/Workspace and Source Generated Documents.md

```diff
@@ -0,0 +1,73 @@
+
+The only surprise you might run into is taking a `SyntaxTree` from an earlier `Solution`, passing it to a newer `Solution`, and getting a document that no longer exists because in the later `Solution` is no longer generating this document. However, passing a `SyntaxTree` between snapshots like that is questionable in the first place. Any code that is working with multiple snapshots and knew that a Document between the snapshots had the same tree could have done something like this, but that code is potentially broken _anyways_ with source generators since now one tree can change due to another tree changing.
+
+## APIs on `Workspace`
+
+### Existing APIs: `OpenDocument(DocumentId)`/`CloseDocument(DocumentId)`
+
+This API today is used by any feature that wants to tell the host to open a file. This will accept the DocumentId of a generated document and work properly.
+
+### Existing APIs: `IsDocumentOpen(DocumentId)` / `GetOpenDocumentIds(...)`
+These will behave no differently than before.
+
+### Existing APIs: `OnDocumentOpened`/`OnDocumentClosed`
+These APIs today associate a Workspace and a SourceTextContainer. Besides allowing APIs like `GetOpenDocumentInCurrentContextWithChanges` to work, this also ensures that a change to the text buffer updates the Workspace automatically. For generated documents, it will wire up the first part (assocating the container) but will not update the Workspace contents when the buffer changes. This is because the updates flow in the _other_ direction, and for now that updating is being managed by a Visual Studio-layer component. Further refactoring may move the core updating into the Workspace layer directly, but not for now.
+
+## APIs for Fetching Documents
+
+### Existing API: `GetOpenDocumentInCurrentContextWithChanges(ITextSnapshot)`
+
+Because we want source generated documents to have the same IDE affordances as any other open file, this API will still work but return a `SourceGeneratedDocument` in that case. Some special handling is required though due to asynchrony. If a user has a generated document open, we may have an async process running in the background that may be trying to refresh this open generated document. The call to `GetOpenDocumentInCurrentContextWithChanges` however will "freeze" the generated document to match the `ITextSnapshot`, so any calls to `GetSourceGeneratedDocumentsAsync()` will return that exact same text, even if that content is out of sync with what would be generated given that project state.
+
+This does mean that in this case, the contents of this document won't match if you compared the document to Workspace.CurrentSolution, got the generated document by ID, and then asked for it's text. This however is OK: that can _always_ be the case for any caller to `GetOpenDocumentInCurrentContextWithChanges` since it always forks to match the `ITextSnapshot`. We're just making the window where this can happen to be bigger than usual.
```

Comment:
not sure i 100% get these areas.  might be worthwhile to give a brief in-person explanation.